# Copyright (c) 2006, 2009, Oracle and/or its affiliates. All rights reserved.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice (including the next
# paragraph) shall be included in all copies or substantial portions of the
# Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.
Raise hsync and default to "1024x768" when DDC data not available.
diff -urp -x '*~' -x '*.orig' hw/xfree86/common/xf86Mode.c hw/xfree86/common/xf86Mode.c
--- hw/xfree86/common/xf86Mode.c	2009-11-22 16:33:40.000000000 -0800
+++ hw/xfree86/common/xf86Mode.c	2009-12-02 18:58:13.140105361 -0800
@@ -1199,6 +1199,7 @@ xf86ValidateModes(ScrnInfoPtr scrp, Disp
     range hsync[MAX_HSYNC];
     range vrefresh[MAX_VREFRESH];
     Bool inferred_virtual = FALSE;
+    Bool UseDefaultRate = FALSE;
 
     DebugF("xf86ValidateModes(%p, %p, %p, %p,\n\t\t  %p, %d, %d, %d, %d, %d, %d, %d, %d, 0x%x)\n",
 	   scrp, availModes, modeNames, clockRanges,
@@ -1244,8 +1245,9 @@ xf86ValidateModes(ScrnInfoPtr scrp, Disp
 		}
 	    } else {
 		scrp->monitor->hsync[0].lo = 31.5;
-		scrp->monitor->hsync[0].hi = 37.9;
+		scrp->monitor->hsync[0].hi = 60.0;
 		scrp->monitor->nHsync = 1;
+		UseDefaultRate = TRUE;
 	    }
 	    type = "default ";
 	}
@@ -1460,6 +1462,11 @@ xf86ValidateModes(ScrnInfoPtr scrp, Disp
 	xf86DeleteMode(&scrp->modes, scrp->modes);
     endp = &scrp->modes;
     last = NULL;
+    if (UseDefaultRate && ((modeNames == NULL) || (*modeNames == NULL))) {
+	modeNames = xnfcalloc(2, sizeof(char *));
+	modeNames[0] = strdup("1024x768");
+	modeNames[1] = NULL;
+    }
     if (modeNames != NULL) {
 	for (i = 0; modeNames[i] != NULL; i++) {
 	    userModes = TRUE;
